# More on Numbers

```{note}
This page is adpted from [Brett, Rush, Adhikari, Denero and Wagner (2023)](https://odsti.github.io/cfd-textbook/Numbers.html). See the [license](https://github.com/pxr687/Intro_Comp_Programming/blob/main/license.md) page for more information.
```

This page will go into more detail about *numbers*, and things every programmer needs to know about them...

Computers are very good calculators. However, there are important things that every programmer working with numbers needs to know. Python (and most other programming languages) uses two *different types of numbers*. We have seen both of them on the [Data Types](1_08_00_Data_Types) page:

* Integers are called `int` values in the Python language. They can only
  represent whole numbers (negative, zero, or positive) that do not have a decimal place. E.g. `1`, `100`, `0` and `-1` are all integers.
  
* Numbers with decimal places are called `float` values (or *floating point values*) in the
  Python language. They can represent whole numbers or numbers with decimal placees. E.g. `10.6`, `1000.0292`, `-17.5` and $\frac{90}{100}$ would all be `float` values in Python, because they all have decimal places. *Note*: decimal places are also called a *fractional component*.

Here are some `int` values:

In [1]:
5

5

In [2]:
11 + 100

111

In [3]:
-3

-3

In [4]:
-25

-25

In [5]:
-1234567890000000000

-1234567890000000000

Here are some float values:

In [6]:
0.2

0.2

In [7]:
15.5 + 20

35.5

In [8]:
30 / 11

2.727272727272727

In [9]:
-12345678900000000000.0

-1.23456789e+19

Look at the output of the last cell. This number is very far from 0 in the negative direction. This means it is a very, very *small* number. Very large or very small `float` values can be displayed using scientific notation. The number you see in the output of the cell above (`-1.23456789e+19`) is shown in scientific notation. Let's look at how to read this number...

## Scientific notation - another way of writing float values

`float` values in scientific notation are in *exponential* format. We can make a literal in exponential format like this:

In [10]:
# Two thousand one hundred.
2.1e3

2100.0

As mentioned above, this exponential format is called *scientific notation*.

`2.1e3` is written in scientific notation, and it is the same value as:

In [11]:
2100.0

2100.0

You can use the `format()` function to make Python show you the number NOT in scientific notation. This is shown in the cell below:

In [12]:
# Take a number in scientific notation, and show it NOT in scientific notation.
format(2.1e3, "f")

'2100.000000'

The `"f"` in the `format()` function tells it that we want [*fixed-point notation*](https://docs.python.org/3/library/string.html). You can think of this as just meaning "NOT scientific notation"! **Note**: this will only show you 6 decimal places, unless you change other settings...(e.g. you can change `"f"` to `".24f"` to show 24 decimal places...).

In [13]:
# Make Python show us 24 decimal places.
format(2.1e3, ".24f")

'2100.000000000000000000000000'

Also note that `format()` returns your number as text/string data, not as `float` data - you will get errors if you ignore this!

Ok, back to scientific notation. 

You can read `2.1e3` as $2.1 * 10^3$.  The `e3` means "times 10 to the power of 3".

So we say `2.1e3` as "2.1 times 10 to the power of 3".

We can also use negative powers to represent small numbers:

In [14]:
1e-3

0.001

Read `1e-3` as $1 x 10^-3$, or "1, but move the decimal point 3 spaces to the
left", or:

In [15]:
1 * 10 ** -3

0.001

How this works is easiest to see as a pattern. Each cell below will show $1 x 10^x$, where $x$ changes by 1:

In [16]:
# 1 multiplied by 10 to the power of 1.
1 * 10 ** 1

10

In [17]:
# 1 multiplied by 10 to the power of 0.
1 * 10 ** 0

1

In [18]:
# 1 multiplied by 10 to the power of -1.
1 * 10 ** -1

0.1

In [19]:
# 1 multiplied by 10 to the power of -2.
1 * 10 ** -2

0.01

In [20]:
# 1 multiplied by 10 to the power of -3.
1 * 10 ** -3

0.001

In [21]:
# 1 multiplied by 10 to the power of -4.
1 * 10 ** -4

0.0001

Look at the output from these cells and make sure you can see the pattern. As we change the exponent (the $x$ in $1 x 10^x$) by -1, our first number "moves" one decimal place to the right. 

Moving the exponent "up" by 1, makes our number go the other way:

| $x$ | $1 * 10^{x}$           | Python Scientific Notation |
| ---: | --------------------: | -------------------------: |
|    6 |       1000000 | `1e06` |
|    5 |       100000 | `1e05` |
|    4 |       10000 | `1e04` |
|    3 |       1000 | `1e03` |
|    2 |       100 | `1e02` |
|    1 |       10 | `1e01` |
|    0 |       1 | `1e00` |
|   -1 | 0.1 | `1e-01` |
|   -2 | 0.01 | `1e-02` |
|   -3 | 0.001 | `1e-03` |
|   -4 | 0.0001 | `1e-04` |
|   -5 | 0.00001 | `1e-05` |
|   -6 | 0.000001 | `1e-06` |

We have used the number 1 to "move around" to show you scientific notation, but it works for any number. So the table above shows:

- 1 times 10 to the power of 6

- 1 times 10 to the power of 5

....

- 1 times 10 to the power of minus 5

- 1 times 10 to the power of minus 6

...but we can change the "1" to any number we want (e.g. "6.23 times 10 to the power of 3" etc.).

Change the value of the variable `try_it_with` in the cell below, to use any number you want. This will help you to get a feel of the meaning of scientific notation: 

**NOTE**: this will only work if you are running the notebook on Binder or Live code. It looks *ugly* on the online textbook! But if you run the page on Binder or Live Code it will look nice, I promise 🙂

In [22]:
# Change this number yourself. 
try_it_with = 4.2

# DO NOT CHANGE THE CODE BELOW THIS POINT.
from mk_scientific_notation_table import scientific_notation_table
scientific_notation_table(try_it_with)

### $\text{Value} = 4.2 * 10^{x}$


| $x$ | Value                  | Python Scientific Notation |
| ---: | --------------------: | -------------------------: |
|    6 |       4200000 | `4.2e06` |
|    5 |       420000 | `4.2e05` |
|    4 |       42000 | `4.2e04` |
|    3 |       4200 | `4.2e03` |
|    2 |       420 | `4.2e02` |
|    1 |       42 | `4.2e01` |
|    0 |       4.2 | `4.2e00` |
|   -1 | 0.42 | `4.2e-01` |
|   -2 | 0.042 | `4.2e-02` |
|   -3 | 0.0042 | `4.2e-03` |
|   -4 | 0.00042 | `4.2e-04` |
|   -5 | 0.000042 | `4.2e-05` |
|   -6 | 0.000004 | `4.2e-06` |


Scientific notation makes it easier write and display very large and very small
numbers. For example, we can write 10 million as:

In [23]:
# 10 million
1e7

10000000.0

Or 0.0000001 as:

In [24]:
# 0.0000001
1e-7

1e-07

This is exactly the same value as:

In [25]:
0.0000001

1e-07

## The limits of float values

> Limit, *noun*, ["the greatest or smallest amount of something that is allowed."](https://www.oxfordlearnersdictionaries.com/definition/english/limit_1)

Float values are very flexible, but they do have limits.

1. A `float` can represent very large and very small numbers. There
   are limits, but you will rarely encounter them.
2. A `float` only represents 15 or 16 significant digits for any number; the
   remaining digits are lost. This limited precision is enough for most uses.
3. After putting together `float` values with mathematical operators, the last few digits may be
   incorrect. Small rounding errors are often confusing when first
   encountered.

The first limit can be seen in two ways. If the result of a calculation is
a very large number, then it is represented as infinite. If the result is a
very small number, then it is represented as zero.

In [26]:
# Very big, but can still be represented by the computer.
2e306 * 10

2e+307

In [27]:
# Too big! Comes back as `inf` for "infinite"
2e306 * 100

inf

In [28]:
# Very small, but can still be represented by the computer.
2e-322 / 10

2e-323

In [29]:
# Too small! Comes back as 0.
2e-322 / 100

0.0

The second limit can be seen by an expression that involves numbers with
more than 15 significant digits. These extra digits are thrown away before any
arithmetic is carried out.

In [30]:
# Our very small digits are NOT lost.
0.666666 - 0.666666123456789

-1.2345678901404256e-07

In [31]:
# Our very small digits ARE lost :(
0.6666666666666666 - 0.6666666666666666123456789 # The digits on the end of the second number are thrown away.

0.0

The third limit can be seen when calculating the difference between two
expressions that should be equivalent. For example, the expression `2 ** 0.5`
computes the square root of 2, but squaring this value does not exactly
make 2. 

Here we `import` the function `math.sqrt()`, which returns the square root of numbers it is given as arguments.

In [32]:
import math
math.sqrt(2)

1.4142135623730951

Remember that the square root of 2 is the number that we can multiply by itself to get 2 e.g. 

$\text{square root of 2} * \text{square root of 2} = 2$

In [33]:
# The square root of 2∗square root of 2
1.4142135623730951 * 1.4142135623730951

2.0000000000000004

In [34]:
# The square root of 2∗square root of 2
math.sqrt(2) * math.sqrt(2)

2.0000000000000004

Let's subtract `2` to get that tiny fractional component by itself:

In [35]:
# The the tiny number by itself.
math.sqrt(2) * math.sqrt(2) - 2

4.440892098500626e-16

The final result above is `0.0000000000000004440892098500626`, a number that
is very close to zero. The correct answer to this arithmetic expression is 0,
but a small error in the final significant digit appears very different in
scientific notation.

This behavior appears in almost all programming languages
because it is the result of the standard way that arithmetic is done by
computers. This is called **floating point error** e.g. very small errors which
happen because of floating point numbers.

Although `float` values are not always exact, they are predictable and work
in the same way across all different kinds of computers and programming
languages.

In summary: 

- doing calculations with very big or very small numbers can lead to floating point errors.

## Before you do the exercise

Make sure you understand this (it can be confusing, so read it carefully):

Think about `4.2e+05`. This base number (`4.2`) is positive, and it has a plus sign `+` in it's scientific notation (e.g. `e+05`). This is a *positive number far from 0*:

In [36]:
4.2e+05

420000.0

However, the number `-4.2e+05` number is *negative*, it has a minus sign `-` at it's start. But it has a plus sign `+` in it's scientific notation (e.g. `e+05`). This is a *negative number far from 0*:

In [37]:
format(-4.2e+05, "f")

'-420000.000000'

The number `4.2e-05` is positive, it does NOT have a minus sign at it's start. BUT it does have a minus sign in its scientific notation (e.g. `e-05`). So, this is a *positive number close to 0*:

In [38]:
format(4.2e-05, ".24f")

'0.000041999999999999997676'

In contrast, the number `-4.2e-05` has a minus sign `-` at it's start, AND in it's scientific notation (e.g. `e-05`). This is a *negative number close to 0*:

In [39]:
format(-4.2e-05, ".24f")

'-0.000041999999999999997676'

So:

| Base Number Sign | `e` Sign | Meaning | Example |
|---|---|---|---|
| Positive (`+`) | Positive (`+`) | **Positive number far from 0** | `4.2e+05` = `420000` |
| Negative (`-`) | Positive (`+`) | **Negative number far from 0** | `-4.2e+05` = `-420000` |
| Positive (`+`) | Negative (`-`) | **Positive number close to 0** | `4.2e-05` = `0.000042` |
| Negative (`-`) | Negative (`-`) | **Negative number close to 0** | `-4.2e-05` = `-0.000042` |

If `e` is POSITIVE (+)  →  move decimal RIGHT  →  number is farther from 0

If `e` is NEGATIVE (-)  →  move decimal LEFT   →  number is closer to 0

The sign of the base number determines which side of 0 it is on, look at the number line below. The `X`'s show the location of different numbers, along with their scientific notation:

```
                                               NEGATIVE                     POSITIVE
                     ◄────────────────────────────────────────── 0 ──────────────────────────────────────────►
Location:                     X                                 X X                                X         
Number:                   -420000                        -0.000042 0.000042                     420000
Scientific notation:      -4.2e+05                       -4.2e-05  4.2e-05                      4.2e+05
Distance from 0:          FAR FROM 0                        CLOSE TO 0                         FAR FROM 0

                   
```

You will need to understand this to do the exercise, so read this section a few times and make sure you understand it. Don't worry if it is confusing at first...

## Exercise - Scientific Notation

We will now do another exercise!

*Remember*: your [assignments, tests and exams](https://pxr687.github.io/Intro_Comp_Programming/1_02_Class_Grades_Assignments_Exams.html) will also be a bit like this.

Click [here 🚀](https://mybinder.org/v2/gh/pxr687/Intro_Comp_Programming/main?urlpath=tree/exercise_2b_scientific_notation.ipynb) to launch the exercise on Binder.

----------------------------------------------------------

[🏡](intro.md).